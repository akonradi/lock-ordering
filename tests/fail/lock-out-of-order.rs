use lock_ordering::{relation::LockAfter, LockLevel, LockedAt, MutualExclusion, Unlocked};

struct FirstLock;
struct SecondLock;

// Either lock can be acquired without acquiring the other, but if the second
// lock *is* held, the first lock can't be acquired.
impl LockAfter<Unlocked> for FirstLock {}
impl LockAfter<Unlocked> for SecondLock {}
impl LockAfter<FirstLock> for SecondLock {}

impl LockLevel for FirstLock {
    type Method = MutualExclusion;
}
impl LockLevel for SecondLock {
    type Method = MutualExclusion;
}

fn main() {
    let first = std::sync::Mutex::new(1234);
    let second = std::sync::Mutex::new('b');

    let mut locked = LockedAt::new();

    // This is fine: the second lock can be acquired without holding the first.
    let (mut locked, mut second_guard) = locked.with_lock::<SecondLock, _>(&second).unwrap();
    *second_guard = 'c';

    // This is problematic: the first lock can't be acquired while the second is
    // held.
    let mut first_guard = locked.lock::<FirstLock, _>(&first);
}
