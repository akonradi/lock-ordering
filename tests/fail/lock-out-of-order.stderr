error[E0277]: the trait bound `FirstLock: LockAfter<SecondLock>` is not satisfied
  --> tests/fail/lock-out-of-order.rs:40:34
   |
40 |     let mut first_guard = locked.lock::<FirstLock>(&first);
   |                                  ^^^^ the trait `LockAfter<SecondLock>` is not implemented for `FirstLock`
   |
   = help: the trait `LockAfter<Unlocked>` is implemented for `FirstLock`
   = help: for that trait implementation, expected `Unlocked`, found `SecondLock`
note: required by a bound in `LockedAt::<'_, L>::lock`
  --> src/lockedat.rs
   |
   |     pub fn lock<'a, NewLock: LockAfter<L> + 'a + MutexLockLevel>(
   |                              ^^^^^^^^^^^^ required by this bound in `LockedAt::<'_, L>::lock`
